### PR TITLE
feat: go-libp2p 0.36 and /webrtc-direct listener

### LINKED ISSUE #9724: Enable WebRTC Transport
# Done criteria
1. libp2p's WebRTC transport is enabled by default.
2. There are corresponding docs that describe how/why to use this.

# Background
WebRTC provides another way for browsers to connect with the rest of the libp2p network (beyond WebTransport which is Chromium only at least as of 2023-03-14).  We want to enable browser nodes to be able to communicate easily with other nodes in the network.  To accomplish this we need a good amount of nodes in the network to have webtransport/webrtc listen addresses.  Getting Kubo nodes having WebTransport listening by default is a good way to increase the number of nodes that browser nodes can connect to.

# Preconditions
1. go-libp2p WebRTC transport (https://github.com/libp2p/go-libp2p/pull/2337 ) is merged and released.

# Notes about the work involved
- [ ] Changelog entry
- [ ] Config update (similar to WebTransport in https://github.com/ipfs/kubo/blob/master/docs/config.md#swarmtransportsnetworkwebtransport )

1. Documenting the browser/Kubo communication story is covered https://github.com/ipfs/ipfs-docs/issues/1286
2. Similar issues from when we did this for WebTransport:
  - https://github.com/ipfs/kubo/pull/9293
  - https://github.com/ipfs/kubo/issues/9292
  - https://github.com/ipfs/kubo/pull/9314
3. With WebTransport we first had this as experimental.  This may change, but my current proposal is that we enable it by default from the start

### LINKED ISSUE #10416: websocket: failed to close network connection: close tcp
### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.tech](https://discuss.ipfs.tech/c/help/13).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/kubo/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [kubo version](https://dist.ipfs.tech/#kubo) or have an issue updating.

### Installation method

ipfs-desktop

### Version

_No response_

### Config

_No response_

### Description

```
[a](http://23.95.11.66:3000/projects/a)
/

ipfs

APP
CPU 0.4%
Memory 410.8 MB
Network I/O 244.1 MB / 158.8 MB
Edit to pin a message for this service
initializing IPFS node at /data/ipfs
Initializing daemon...
Kubo version: 0.28.0-e7f0f34
Repo version: 15
System version: amd64/linux
Golang version: go1.22.2
2024/05/04 18:58:32 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/quic-go/quic-go/wiki/UDP-Buffer-Sizes for details.
Swarm listening on /ip4/10.0.1.174/tcp/4001
Swarm listening on /ip4/10.0.1.174/udp/4001/quic-v1
Swarm listening on /ip4/10.0.1.174/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /ip4/10.0.2.129/tcp/4001
Swarm listening on /ip4/10.0.2.129/udp/4001/quic-v1
Swarm listening on /ip4/10.0.2.129/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip4/127.0.0.1/udp/4001/quic-v1
Swarm listening on /ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /ip4/172.19.0.20/tcp/4001
Swarm listening on /ip4/172.19.0.20/udp/4001/quic-v1
Swarm listening on /ip4/172.19.0.20/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /p2p-circuit
Swarm announcing /ip4/127.0.0.1/tcp/4001
Swarm announcing /ip4/127.0.0.1/udp/4001/quic-v1
Swarm announcing /ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm announcing /ip4/172.19.0.20/tcp/4001
Swarm announcing /ip4/172.19.0.20/udp/4001/quic-v1
Swarm announcing /ip4/172.19.0.20/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm announcing /ip4/23.95.11.66/udp/32980/quic-v1
Swarm announcing /ip4/23.95.11.66/udp/32980/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
RPC API server listening on /ip4/0.0.0.0/tcp/5001
WebUI: http://0.0.0.0:5001/webui
Gateway server listening on /ip4/0.0.0.0/tcp/8080
Daemon is ready
2024/05/04 18:58:56 websocket: failed to close network connection: close tcp 172.19.0.15:57232->145.40.118.135:443: use of closed network connection
2024/05/04 18:59:38 websocket: failed to close network connection: close tcp 172.19.0.20:58760->145.40.118.135:443: use of closed network connection
2024/05/04 19:00:47 websocket: failed to close network connection: close tcp 172.19.0.20:45834->172.64.144.159:443: use of closed network connection
2024/05/04 19:01:39 websocket: failed to close network connection: close tcp 172.19.0.20:43902->145.40.118.135:443: use of closed network connection
2024/05/04 19:02:04 websocket: failed to close network connection: close tcp 172.19.0.15:35512->145.40.118.135:443: use of closed network connection
2024/05/04 19:02:27 websocket: failed to close network connection: close tcp 172.19.0.20:56978->145.40.118.135:443: use of closed network connection
2024/05/04 19:04:11 websocket: failed to close network connection: close tcp 172.19.0.15:37282->145.40.118.135:443: use of closed network connection
2024/05/04 19:05:23 websocket: failed to close network connection: close tcp 172.19.0.15:33652->145.40.118.135:443: use of closed network connection
Changing user to ipfs
ipfs version 0.28.0
Found IPFS fs-repo at /data/ipfs
Initializing daemon...
Error: lock /data/ipfs/repo.lock: someone else has the lock
Kubo version: 0.28.0-e7f0f34
Repo version: 15
System version: amd64/linux
Golang version: go1.22.2

Use
+
K
to trigger search


Logs
Lines
50
initializing IPFS node at /data/ipfs
Initializing daemon...
Kubo version: 0.28.0-e7f0f34
Repo version: 15
System version: amd64/linux
Golang version: go1.22.2
2024/05/04 18:58:32 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/quic-go/quic-go/wiki/UDP-Buffer-Sizes for details.
Swarm listening on /ip4/10.0.1.174/tcp/4001
Swarm listening on /ip4/10.0.1.174/udp/4001/quic-v1
Swarm listening on /ip4/10.0.1.174/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /ip4/10.0.2.129/tcp/4001
Swarm listening on /ip4/10.0.2.129/udp/4001/quic-v1
Swarm listening on /ip4/10.0.2.129/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip4/127.0.0.1/udp/4001/quic-v1
Swarm listening on /ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /ip4/172.19.0.20/tcp/4001
Swarm listening on /ip4/172.19.0.20/udp/4001/quic-v1
Swarm listening on /ip4/172.19.0.20/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm listening on /p2p-circuit
Swarm announcing /ip4/127.0.0.1/tcp/4001
Swarm announcing /ip4/127.0.0.1/udp/4001/quic-v1
Swarm announcing /ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm announcing /ip4/172.19.0.20/tcp/4001
Swarm announcing /ip4/172.19.0.20/udp/4001/quic-v1
Swarm announcing /ip4/172.19.0.20/udp/4001/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
Swarm announcing /ip4/23.95.11.66/udp/32980/quic-v1
Swarm announcing /ip4/23.95.11.66/udp/32980/quic-v1/webtransport/certhash/uEiBC1lPftzpyh1tTvr_DmNbimm5r3TRhYrS6B3oB6hDInA/certhash/uEiDtaSGsb2bB2cqNMhCWw5rwGwbrK1eA1DAEqYfA5XyguQ
RPC API server listening on /ip4/0.0.0.0/tcp/5001
WebUI: http://0.0.0.0:5001/webui
Gateway server listening on /ip4/0.0.0.0/tcp/8080
Daemon is ready
2024/05/04 18:58:56 websocket: failed to close network connection: close tcp 172.19.0.15:57232->145.40.118.135:443: use of closed network connection
2024/05/04 18:59:38 websocket: failed to close network connection: close tcp 172.19.0.20:58760->145.40.118.135:443: use of closed network connection
2024/05/04 19:00:47 websocket: failed to close network connection: close tcp 172.19.0.20:45834->172.64.144.159:443: use of closed network connection
2024/05/04 19:01:39 websocket: failed to close network connection: close tcp 172.19.0.20:43902->145.40.118.135:443: use of closed network connection
2024/05/04 19:02:04 websocket: failed to close network connection: close tcp 172.19.0.15:35512->145.40.118.135:443: use of closed network connection
2024/05/04 19:02:27 websocket: failed to close network connection: close tcp 172.19.0.20:56978->145.40.118.135:443: use of closed network connection
2024/05/04 19:04:11 websocket: failed to close network connection: close tcp 172.19.0.15:37282->145.40.118.135:443: use of closed network connection
2024/05/04 19:05:23 websocket: failed to close network connection: close tcp 172.19.0.15:33652->145.40.118.135:443: use of closed network connection
Changing user to ipfs
ipfs version 0.28.0
Found IPFS fs-repo at /data/ipfs
Initializing daemon...
Error: lock /data/ipfs/repo.lock: someone else has the lock
Kubo version: 0.28.0-e7f0f34
Repo version: 15
System version: amd64/linux
Golang version: go1.22.2
```

### PR DESCRIPTION
**Pull request by lidel**
This PR enables WebRTC Direct transport and related listener by default.

updated.

- [x] test with https://github.com/libp2p/go-libp2p/pull/2887
- [x] changelog and docs
- [x] add `/udp/4002/webrtc-direct` listener by default (`ipfs init`)
- [x] add `/udp/400X/webrtc-direct` listener for existing users (write migration)
  - some users will have 4002, 4003 taken, migration needs to be smart enough to pick next free port
  - we can use /webtransport one we did in https://github.com/ipfs/fs-repo-migrations/pull/162 as a template
  - [x] https://github.com/ipfs/fs-repo-migrations/pull/190
  - [x] https://github.com/ipfs/distributions/pull/1091 
- [x] https://github.com/ipfs/kubo/pull/10462  should land first, and then this one should be tested on top of it
- [x] go-libp2p allows for listening on the same UDP port as /quic-v1 
  - [x] switch this PR to use same `/udp` port (4001)
- [x] switch to [go-libp2p 0.36](https://github.com/libp2p/go-libp2p/releases/tag/v0.36.0)

Closes #9724
Closes #10416 as well (because go-libp2p now pulls-in gorilla/websocket v1.5.3)
Part of https://github.com/ipfs/kubo/issues/10436

### CODE REVIEW

**Comment by lidel** [comment]
(Marking as blocked until go-libp2p allows listening on the same UDP port as `/quic-v1`)

File: go.mod [hunk 1]
```diff
@@ -48,7 +48,7 @@ require (
 	github.com/jbenet/goprocess v0.1.4
 	github.com/julienschmidt/httprouter v1.3.0
 	github.com/libp2p/go-doh-resolver v0.4.0
-	github.com/libp2p/go-libp2p v0.35.4
+	github.com/libp2p/go-libp2p v0.35.1-0.20240730155614-111d6df08497
```
Comment (lidel) [inline go.mod:51]: @MarcoPolo I got below `fx` error ([log](https://github.com/ipfs/kubo/actions/runs/10167054881/job/28118558695#step:5:8645)) in after switching to `111d6df08497` from https://github.com/libp2p/go-libp2p/pull/2889:

`failed to build host.Host: received non-nil error from function "github.com/ipfs/kubo/core/node/libp2p".Host (github.com/ipfs/kubo/core/node/libp2p/host.go:41): could not build arguments for function "reflect".makeFuncStub (reflect/asm_amd64.s:28): could not build value group transport.Transport[group="transport"]: missing dependencies for function "reflect".makeFuncStub (reflect/asm_amd64.s:28): missing type: libp2pwebrtc.ListenUDPFn`

Is this something obvious, or still wip?

File: go.sum [hunk 2]
```diff
@@ -330,8 +330,8 @@ github.com/ipfs-shipyard/nopfs/ipfs v0.13.2-0.20231027223058-cde3b5ba964c h1:7Uy
 github.com/ipfs-shipyard/nopfs/ipfs v0.13.2-0.20231027223058-cde3b5ba964c/go.mod h1:6EekK/jo+TynwSE/ZOiOJd4eEvRXoavEC3vquKtv4yI=
 github.com/ipfs/bbloom v0.0.4 h1:Gi+8EGJ2y5qiD5FbsbpX/TMNcJw8gSqr7eyjHa4Fhvs=
 github.com/ipfs/bbloom v0.0.4/go.mod h1:cS9YprKXpoZ9lT0n/Mw/a6/aFV6DTjTLYHeA+gyqMG0=
-github.com/ipfs/boxo v0.21.1-0.20240730141506-8e51658f8a1b h1:vH+ly+ghScnm4fDTsFnRkWIqZ0Vux6U3aOzPc3bQsOM=
-github.com/ipfs/boxo v0.21.1-0.20240730141506-8e51658f8a1b/go.mod h1:NmweAYeY1USOaJJxouy7DLr/Y5M8UBSsCI2KRivO+TY=
+github.com/ipfs/boxo v0.20.0 h1:umUl7q1v5g5AX8FPLTnZBvvagLmT+V0Tt61EigP81ec=
+github.com/ipfs/boxo v0.20.0/go.mod h1:mwttn53Eibgska2DhVIj7ln3UViq7MVHRxOMb+ehSDM=
```
Comment (lidel) [inline go.sum:334]: :thought_balloon: unsure why go mode downgraded things here, likely boxo uses more modern deps than go-libp2p

File: go.mod (hunk 1 above)
Comment (lidel) [reply go.mod:51]: fwiw confirmed code from https://github.com/libp2p/go-libp2p/pull/2889 is used and loads fine. Can create objects that implement `libp2pwebrtc.ListenUDPFn`, and still get this error, so this looks like a bug specific to how `fx`  works. 

I got the same error when in Rainbow.

Comment (MarcoPolo) [reply go.mod:51]: My apologies, I forgot to add the provide in a certain code path for the autonat host. I just pushed a commit that fixes this.

Comment (lidel) [reply go.mod]: @MarcoPolo I'm OoO today, but checked the version from https://github.com/ipfs/kubo/pull/10463/commits/6513aedde13403c37d91d0756e00ee753f8c21e5 before going AFK, and seems and it seems to work fine, listener on the same port initialized correctly and I was able to run libp2p identify over `/udp/4001/quic-v1`, so no regression on existing transports, which is the most important.

I'll be back Thu but feel free to move forward with https://github.com/libp2p/go-libp2p/pull/2889.

(CI fails for unrelated reason (rhis PR downgraded boxo version)).

**Comment by lidel** [comment]
Confirmed:
- [x]  `/udp/4001/webrtc-direct` multiaddr works on  https://helia-identify.on.fleek.co/
- [x] Confirmed public `/udp/4001/webrtc-direct` listener is correctly inferred from the public IP learned via AutoNAT and other means (private LAN with UPnP). 

I'll now work on migration to add listener to existing user's machines.

**Comment by lidel** [comment]
Migrations are ready (fs-repo-15-to-16 work happened in: https://github.com/ipfs/fs-repo-migrations/pull/190, https://github.com/ipfs/distributions/pull/1091).

One go-test fails due to boxo version being lowered – this PR is waiting for final go-libp2p release (which hopefully will resolve dep conflict with boxo).

**Review by lidel — COMMENTED** [review]
Merging to have wider window of testing `master` with go-libp2p 0.36.1 before we tag Kubo 0.30.0-rc1, and also rebase #10468

**Pull request merged**